pkg/metrics/v2/scraper.go

```diff
@@ -0,0 +1,318 @@
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	pb "github.com/grafana/agent/pkg/metrics/v2/internal/metricspb"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	prom_config "github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/scrape"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type scraperMetrics struct {
+	util.MetricsContainer
+
+	numberScrapers    prometheus.Gauge
+	totalTargetPushes prometheus.Counter
+	totalTargets      prometheus.Counter
+	totalFailedPushes *prometheus.CounterVec
+
+	scraperTargets *prometheus.GaugeVec
+}
+
+func newScraperMetrics() *scraperMetrics {
+	var m scraperMetrics
+
+	m.numberScrapers = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "agent_metrics_scrapers_count",
+		Help: "Current number of running scrapers",
+	})
+	m.totalTargetPushes = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "agent_metrics_scrapers_received_pushes_total",
+		Help: "Number of times this node has received a set of targets from other nodes",
+	})
+	m.totalTargets = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "agent_metrics_scrapers_received_targets_total",
+		Help: "Total number of targets this node has received from other nodes",
+	})
+	m.totalFailedPushes = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "agent_metrics_scrapers_failed_pushes_total",
+		Help: "Number of times scrapers failed to receieve targets",
+	}, []string{"reason"})
+
+	m.scraperTargets = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "agent_metrics_scraper_targets",
+		Help: "Current number of targets for the scraper",
+	}, []string{"instance"})
+
+	m.Add(
+		m.numberScrapers,
+		m.totalTargetPushes,
+		m.totalTargets,
+		m.totalFailedPushes,
+
+		m.scraperTargets,
+	)
+	return &m
+}
+
+// scraperManager implements metricspb.ScraperServer and will manage scrapers
+// based on receieved targets.
+type scraperManager struct {
+	pb.UnimplementedScraperServer
+
+	log     log.Logger
+	ss      storageSet
+	metrics *scraperMetrics
+
+	mut              sync.RWMutex
+	stopped          bool
+	instanceGroups   map[string]targetGroups // Map of instance name -> target groups
+	scraperInstances map[string]*scraper     // Running scrapers.
+}
+
+type targetGroups = map[string][]*targetgroup.Group
+
+// newScraperManager creates a new scraperManager. No scapers are available
+// until ApplyConfig is called.
+func newScraperManager(log log.Logger, ss storageSet) *scraperManager {
+	return &scraperManager{
+		log:     log,
+		ss:      ss,
+		metrics: newScraperMetrics(),
+
+		instanceGroups:   make(map[string]targetGroups),
+		scraperInstances: make(map[string]*scraper),
+	}
+}
+
+// ApplyConfig will update the managed set of scrapers. ApplyConfig should not
+// be called util the storageSet is reconfigured first.
+func (sm *scraperManager) ApplyConfig(cfg *Config) error {
```

Comment:
This should detect if a scrape job goes away, since the Discoverer won't send an empty target set for scrape jobs it doesn't know about.